src/validation/rules/no_unused_variables.rs

```diff
@@ -0,0 +1,370 @@
+use std::collections::HashSet;
+
+use super::ValidationRule;
+use crate::ast::AstNodeWithName;
+use crate::static_graphql::query::{OperationDefinition, VariableDefinition};
+use crate::validation::utils::{ValidationError, ValidationErrorContext};
+use crate::{
+    ast::{ext::AstWithVariables, QueryVisitor},
+    validation::utils::ValidationContext,
+};
+
+/// No unused fragments
+///
+/// A GraphQL operation is only valid if all variables defined by an operation
+/// are used, either directly or within a spread fragment.
+///
+/// See https://spec.graphql.org/draft/#sec-All-Variables-Used
+pub struct NoUnusedVariables;
+
+struct NoUnusedVariablesHelper<'a> {
+    variables_definitions: HashSet<String>,
+    error_context: ValidationErrorContext<'a>,
+}
+
+impl<'a> NoUnusedVariablesHelper<'a> {
+    fn new(validation_context: &'a ValidationContext<'a>) -> Self {
+        NoUnusedVariablesHelper {
+            variables_definitions: HashSet::new(),
+            error_context: ValidationErrorContext::new(validation_context),
+        }
+    }
+}
+
+impl<'a> QueryVisitor<NoUnusedVariablesHelper<'a>> for NoUnusedVariables {
+    fn enter_variable_definition(
+        &self,
+        node: &VariableDefinition,
+        _parent_operation: &OperationDefinition,
+        visitor_context: &mut NoUnusedVariablesHelper<'a>,
+    ) {
+        visitor_context
+            .variables_definitions
+            .insert(node.name.clone());
+    }
+
+    fn enter_operation_definition(
+        &self,
+        _node: &OperationDefinition,
+        visitor_context: &mut NoUnusedVariablesHelper<'a>,
+    ) {
+        visitor_context.variables_definitions.clear();
+    }
```

Comment:
You can remove the entire section, and also the `visitor_context.variables_definitions`, and do inside `leave_operation_definition`:  `let variables = operation.get_variables()`